File: datalib/src/main/java/nl/acidcats/tumblrlikes/core/usecases/photos/UpdatePhotoCacheUseCaseImpl.java
package nl.acidcats.tumblrlikes.core.usecases.photos;

import java.util.List;

import javax.inject.Inject;

import nl.acidcats.tumblrlikes.core.models.Photo;
import nl.acidcats.tumblrlikes.core.repositories.PhotoDataRepository;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created on 01/10/2018.
 */
public class UpdatePhotoCacheUseCaseImpl implements UpdatePhotoCacheUseCase {

    private PhotoDataRepository _photoDataRepository;

    @Inject
    UpdatePhotoCacheUseCaseImpl(PhotoDataRepository photoDataRepository) {
        _photoDataRepository = photoDataRepository;
    }

    @Override
    public Observable<Boolean> removeCachedHiddenPhotos() {
        return _photoDataRepository
                .removeCachedHiddenPhotos()
                .observeOn(AndroidSchedulers.mainThread());
    }

    @Override
    public Observable<Integer> checkCachedPhotos() {
        return Observable
                .fromCallable(() -> _photoDataRepository.getCachedPhotos())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .flatMapIterable(photoEntities -> photoEntities)
                .filter(photo -> _photoDataRepository.isPhotoCacheMissing(photo))
                .map(Photo::getId)
                .toList()
                .map(ids -> _photoDataRepository.setPhotosUncached(ids))
                .map(List::size);
    }
}

File: datalib/src/main/kotlin/nl/acidcats/tumblrlikes/core/usecases/photos/UpdatePhotoCacheUseCaseImpl.kt
package nl.acidcats.tumblrlikes.core.usecases.photos

import nl.acidcats.tumblrlikes.core.models.Photo
import nl.acidcats.tumblrlikes.core.repositories.PhotoDataRepository
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created on 24/10/2018.
 */
class UpdatePhotoCacheUseCaseImpl @Inject constructor(private val photoDataRepository: PhotoDataRepository) : UpdatePhotoCacheUseCase {

    override fun removeCachedHiddenPhotos(): Observable<Boolean> {
        return photoDataRepository
                .removeCachedHiddenPhotos()
                .observeOn(AndroidSchedulers.mainThread())
    }

    // TODO use filter & map from collections to do filtering & mapping
    override fun checkCachedPhotos(): Observable<Int> {
        return Observable
                .fromCallable { photoDataRepository.getCachedPhotos() }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .flatMapIterable { it }
                .filter(photoDataRepository::isPhotoCacheMissing)
                .map(Photo::id)
                .toList()
                .map(photoDataRepository::setPhotosUncached)
                .map(List<Long>::size)
    }
}